src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs

```diff
@@ -164,6 +168,17 @@ bool HasMoreToSync(out BlockHeader[]? headers, out int headersToRequest)
                 }
 
                 if (cancellation.IsCancellationRequested) return blocksSynced; // check before every heavy operation
+
+                if (shouldProcess)
+                {
+                    BlockHeader? parentHeader = _blockTree.FindHeader(headers[0].ParentHash!);
+                    if (parentHeader is not null && !_stateReader.HasStateForBlock(parentHeader))
```

Comment:
Can't we do it with full state finder and just check if the full state number is < parent header? I just don't like unnecessary change.

Reply:
So we need to check for every block looks like

Reply:
I guess would be perfect to track what state is being downloaded and download blocks without processing till that block + don't download if potential state is not known